libdrgn/python/module.c

```diff
@@ -10,6 +10,52 @@ PyObject *FaultError;
 PyObject *MissingDebugInfoError;
 PyObject *OutOfBoundsError;
 
+int FaultError_init(PyObject *self, PyObject *args, PyObject *kwds)
+{
+	PyObject *address, *message;
+
+	if (!_PyArg_NoKeywords(Py_TYPE(self)->tp_name, kwds))
+		return -1;
+
+	if (!PyArg_ParseTuple(args, "OO:FaultError", &message, &address))
+		return -1;
+
+	if (PyObject_SetAttrString(self, "message", message) < 0)
+		return -1;
+
+	if (PyObject_SetAttrString(self, "address", address) < 0)
+		return -1;
+
+	return 0;
+}
+
+static PyObject *FaultError_str(PyObject *self)
+{
+	PyObject *message, *address;
+
+	message = PyObject_GetAttrString(self, "message");
+	if (!message)
+		return NULL;
+
+	address = PyObject_GetAttrString(self, "address");
+	if (!address)
+		return NULL;
+
+	return PyUnicode_FromFormat("%S: %S",
+				    message, PyNumber_ToBase(address, 16));
```

Comment:
This is leaking references to the message and address attributes and the return value from `PyNumber_ToBase()`. I'd write it like this:

```c
static PyObject *FaultError_str(PyObject *self)
{
	PyObject *message, *address, *fmt, *ret = NULL;

	message = PyObject_GetAttrString(self, "message");
	if (!message)
		return NULL;

	address = PyObject_GetAttrString(self, "address");
	if (!address)
		goto out_message;

	fmt = PyUnicode_FromString("%s: %#x");
	if (!fmt)
		goto out_address;

	ret = PyUnicode_Format(fmt, message, address);

	Py_DECREF(fmt);
out_address:
	Py_DECREF(address);
out_message:
	Py_DECREF(message);
	return ret;
}
```